pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
-	webhook, err := m.readyWebhookConfiguration()
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to get %s webhook configuration %s", m.webhookType, m.webhookName)
+func mapCASecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	secret.Type = corev1.SecretTypeOpaque
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
 	}
+	secret.Data[CAPrivateKeyKey] = certificateChain.CA.KeyPEM
+	secret.Data[CACertKey] = certificateChain.CA.CertPEM
+}
 
-	clientConfigList := m.clientConfigList(webhook)
-	if clientConfigList == nil || len(clientConfigList) == 0 {
-		return nil, errors.Wrapf(err, "failed to access CABundle clientConfigList is empty in %s webhook configuration %s", m.webhookType, m.webhookName)
+func newObjectKey(kind objectKind, namespace, name string) *objectKey {
+	key := objectKey{
+		Kind: kind,
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespace,
+		},
 	}
+	return &key
+}
+
+func (m *Manager) secretCAName() types.NamespacedName {
+	return types.NamespacedName{m.namespace, m.name + "-ca"}
+}
+
+func caBundleName(webhookName, configName string) string {
+	return webhookName + "-" + configName
+}
+
+func namespacedHostname(name, namespace string) string {
+	return name + "." + namespace
+}
 
-	return clientConfigList[0].CABundle, nil
+func serviceHostname(name, namespace string) string {
+	return name + "." + namespace + serviceSubdomain
 }
 
-// getServicesFromConfiguration it retrieves all the references services at
-// webhook configuration clientConfig and in case there is URL instead of
-// ServiceRef it will reference fake one with webhook name, mgr namespace and
-// passing the url hostname at map value
-func (m *Manager) getServicesFromConfiguration(configuration client.Object) (map[types.NamespacedName][]string, error) {
+func serviceFqdn(name, namespace string) string {
+	return name + "." + namespace + serviceSubdomain + clusterDomain
+}
 
-	services := map[types.NamespacedName][]string{}
+func newCertificateIssue(name, namespace string) *chain.CertificateIssue {
+	commonName := serviceHostname(name, namespace)
+	hostnames := []string{
+		name,
+		namespacedHostname(name, namespace),
+		commonName,
+		serviceFqdn(name, namespace),
+	}
+	certificateBundle := chain.CertificateIssue{
+		Name:      commonName,
+		Hostnames: hostnames,
+		CACertPEM: make(map[string][]byte),
+	}
+	return &certificateBundle
+}
 
-	for _, clientConfig := range m.clientConfigList(configuration) {
+func mutatingWebhookConfig(webhook client.Object) *admissionregistrationv1.MutatingWebhookConfiguration {
+	return webhook.(*admissionregistrationv1.MutatingWebhookConfiguration)
+}
 
-		service := types.NamespacedName{}
-		hostnames := []string{}
+func validatingWebhookConfig(webhook client.Object) *admissionregistrationv1.ValidatingWebhookConfiguration {
+	return webhook.(*admissionregistrationv1.ValidatingWebhookConfiguration)
+}
```

Comment:
Let's remove this helpers there are used in just one place and open the door to future changes using them.